solutions/usaco-815.mdx

```diff
@@ -0,0 +1,98 @@
+---
+id: usaco-815
+source: USACO Gold 2018 Febuary
+title: Taming the Herd
+author: Shijie Ren
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_taming_gold_feb18.html)
+
+## Explanation
+
+Let $\texttt{dp}[i][j]$ be the minimum answer of first $j$ days with $i$
+breakouts.
+
+On the $j$-th day, there's 3 cases:
+
+- if log is tampered and cows would NOT breakout:
+
+$$
+\texttt{dp}[i][j]=\texttt{dp}[i][j-1]+1
+$$
+
+- if log is tampered and cows would breakout (So $a[j] \not = 0$):
+
+$$
+\texttt{dp}[i][j]=\texttt{dp}[i-1][j-1]+1
+$$
+
+- if log is NOT tampered (Then $i$-th breakout will be on $(j-a[j])$-th day):
+
+$$
+\texttt{dp}[i][j] = \texttt{dp}[i-1][j-a[j]-1]+\texttt{range\_ans}[j-a[j]][j]);
+$$
+
+$$\texttt{range\_ans}[l][r]$$ represents minimum cost to replace range
+$a[l]...a[r]$ with $0...(r-l)$ respectively.
+
+So if the state meets the conditions,
+
+$$
+\texttt{dp}[i][j] = \max \{ \texttt{dp}[i][j-1]+1 ,\texttt{dp}[i-1][j-1]+1 ,\texttt{dp}[i-1][j-a[j]-1] + \texttt{range\_ans}[j-a[j]][j]) \}
+$$
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N^2)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <algorithm>
+#include <cstdio>
+#include <cstring>
+#include <iostream>
+#define N 105
+#define byteinf 0x3f
```

Comment:
use `const`
`#define` is cancer
also for inf just use INT32_MAX or INT64_MAX

Reply:
INT32_MAX might cause integer overflow?

Reply:
if that's the case i recommend you just do a couple of simple if statements